Revise "mf_vn" (#2913)
aivivn/d2l-vn: chapter_recommender-systems/mf_vn.md

<!--
Matrix Factorization :cite:`Koren.Bell.Volinsky.2009` is a well-established algorithm in the recommender systems literature.
The first version of matrix factorization model is proposed by Simon Funk in a famous [blog post](https://sifter.org/~simon/journal/20061211.html) 

> in which he described the idea of factorizi
> ng the interaction matrix.

Reviewer: ```suggestion
in which he described the idea of factorizing the interaction matrix.
```

The predicted ratings can be estimated by
-->

Gọi $\mathbf{R} \in \mathbb{R}^{m \times n}$ ký hiệu ma trận tương tác với $m$ người dùng và $n$ sản phẩm, 
và các giá trị $\mathbf{R}$ biểu diễn đánh giá trực tiếp. 
Tương tác người dùng - sản phẩm được phân tích thành ma trận người dùng tiềm ẩn $\mathbf{P} \in \mathbb{R}^{m \times k}$
và ma trận sản phẩm tiềm ẩn $\mathbf{Q} \in \mathbb{R}^{n \times k}$, trong đó $k \ll m, n$, là kích thước nhân tố tiềm ẩn. 
Gọi $\mathbf{p}_u$ ký hiệu hàng thứ $u$ và của $\mathbf{P}$ và $\mathbf{q}_i$ ký hiệu hàng thứ $i$ của $\mathbf{Q}$. 

> Với một sản phẩm $i$ cho trước, các phần tử trong $\mathbf{q}_i$ đo lường mức độ mà sản phẩm sở hữu các đặc trưng ví dụ như thể loại hay ngôn ngữ của một bộ phim.

Reviewer: ```suggestion
Với một sản phẩm $i$ cho trước, các phần tử trong $\mathbf{q}_i$ đo lường mức độ mà sản phẩm đó sở hữu các đặc trưng, ví dụ như thể loại hay ngôn ngữ của một bộ phim.
```

> Với một người dùng $u$ cho trước, các phần tử trong $\mathbf{p}_u$ đo mức độ ưa thích của người dùng này đối với các đặc trưng thuộc sản phẩm tương ứng. 

Reviewer: ```suggestion
Với một người dùng $u$ cho trước, các phần tử trong $\mathbf{p}_u$ đo mức độ ưa thích của người dùng này đối với các đặc trưng tương ứng của các sản phẩm. 
```

Specifically, the predicted rating user $u$ gives to item $i$ is calculated by
-->

trong đó $\hat{\mathbf{R}}\in \mathbb{R}^{m \times n}$ là ma trận đánh giá dự đoán và có cùng kích thước với $\mathbf{R}$. 
Một vấn đề lớn của cách dự đoán này là độ chệch (*bias*) của người dùng/sản phẩm không được mô hình hoá. 
Ví dụ, một số người dùng có thiên hướng đánh giá cao hơn, hoặc một số sản phẩm luôn bị đánh giá thấp hơn bởi chất lượng kém. 

> Các độ chệch này là rất bình thường trong những ứng dụng thực tế. 

Reviewer: ```suggestion
Các độ chệch này là rất phổ biến trong những ứng dụng thực tế. 
```

> Để thu được các độ chệch này, số hạng độ chệch của người dùng và sản phẩm riêng biệt được sử dụng.

Reviewer: ```suggestion
Để thu được các độ chệch này, số hạng độ chệch riêng biệt cho từng người dùng và sản phẩm được sử dụng.
```

In the `forward` function, user and item ids are used to look up the embeddings.
-->

Đầu tiên, ta lập trình mô hình phân rã ma trận như mô tả trên. 
Các nhân tố tiềm ẩn của người dùng và sản phẩm được tạo bằng `nn.Embedding`.
Tham số `input_dim` là số sản phẩm/người dùng và `output_dim` là kích thước nhân tố tiềm ẩn ($k$). 
Ta cũng có thể sử dụng `nn.Embedding` để tạo độ chệch cho người dùng/sản phẩm bằng cách gán `output_dim` bằng một. 

> Trong hàm `forward`, id người dùng và sản phẩm được sử dụng để tìm đến đối tượng embedding. 

Reviewer: ```suggestion
Trong hàm `forward`, id người dùng và sản phẩm được sử dụng để truy vấn tới embedding tương ứng. 
```

Below, we use the trained model to predict the rating that a user (ID 20) might give to an item (ID 30).
-->

> Ở dưới, ta sử dụng mô hình đã được huấn luyện để dự đoán đánh giá có thể của một người dùng (ID 20) cho một sản phẩm (ID 30). 

Reviewer: ```suggestion
Ở dưới, ta sử dụng mô hình đã được huấn luyện để dự đoán đánh giá mà một người dùng (ID 20) có thể gán cho một sản phẩm (ID 30). 
```

<!-- Phần 2 -->
* Đỗ Trường Giang
* Nguyễn Văn Cường

> * Nguyễn Lê Quang Nhật

Author: ```suggestion
* Nguyễn Lê Quang Nhật
* Lê Khắc Hồng Phúc
```
